### PR TITLE
SNOW-2129018 async sproc

### PR DESCRIPTION
**Pull request by sfc-gh-mayliu**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-2129018

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [ ] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://github.com/snowflakedb/snowpark-python/blob/main/CONTRIBUTING.md#thread-safe-development)
   - [ ] If adding any arguments to public Snowpark APIs or creating new public Snowpark APIs, I acknowledge that I have ensured my changes include AST support. Follow the link for more information: [AST Support Guidelines](https://github.com/snowflakedb/snowpark-python/blob/main/CONTRIBUTING.md#ast-abstract-syntax-tree-support-in-snowpark)

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

Adds asynchronous stored procedure execution support. `Session.call()` and `StoredProcedure.__call__()` now accept a `block: bool = True` parameter. When set to False, these methods submit the stored procedure query and immediately return an AsyncJob handle, allowing non-blocking execution. A convenience method `Session.call_nowait()` is also introduced as shorthand for `Session.call(block=False)`.

### CODE REVIEW

**Review by sfc-gh-aalam — APPROVED** [review]
Looks great!

File: src/snowflake/snowpark/session.py [hunk 1]
```diff
@@ -4513,3 +4576,53 @@ def _explain_query(self, query: str) -> Optional[str]:
...
+                results_cursor = self._conn.execute_async_and_notify_query_listener(
+                    query, _statement_params=statement_params
+                )
+                return AsyncJob(
+                    results_cursor["queryId"],
+                    query,
+                    self,
+                    _AsyncResultType.COUNT,
+                    log_on_exception=log_on_exception,
+                )
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/session.py]: nit: I feel there's some redundancy in this part of code, for the async part, as far as I can see the only difference is _AsyncResultType.COUNT vs _AsyncResultType.ROW

e.g.,

the returning can just be the following for async job

```python
                results_cursor = self._conn.execute_async_and_notify_query_listener(
                    query, _statement_params=statement_params
                )
                return AsyncJob(
                    results_cursor["queryId"],
                    query,
                    self,
                    _AsyncResultType. ROW if is_return_table else _AsyncResultType.COUNT,
                    log_on_exception=log_on_exception,
                )
```

File: src/snowflake/snowpark/stored_procedure.py [hunk 2]
```diff
@@ -153,6 +154,15 @@ def __call__(
...
+                    query=query,
+                    is_return_table=self._is_return_table,
+                    block=block,
+                    statement_params=statement_params,
+                    ast_stmt=self._ast_stmt,
+                )
+
             if self._is_return_table:
                 qid = session._conn.execute_and_get_sfqid(
                     query, statement_params=statement_params
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/stored_procedure.py]: is the logic 166-170 encapsulated into `_execute_sproc_internal`? is it possible to do the following?

```python
ret = _execute_sproc_internal(...block=block, return_table=self._is_return_table)
if not block:
    return ret
else:
    ret._ast = sproc_expr
    return ret
```

File: tests/integ/test_stored_procedure.py [hunk 3]
```diff
@@ -1282,6 +1283,297 @@ def select_star_decorator_sp(session_: Session, name: str) -> DataFrame:
...
+            input_types=[IntegerType()],
+            return_type=StringType(),
+            replace=True,
+        )
+
+        # 1a: session.call(block=False)
+        async_job_scalar = session.call(scalar_sp_name, 17, block=False)
+        assert isinstance(async_job_scalar, AsyncJob)
+        while not async_job_scalar.is_done():
+            time.sleep(1)
```
Comment (sfc-gh-aling) [inline tests/integ/test_stored_procedure.py]: do we need the while loop here?
`AsyncJob.result()` blocks until it completes

(same comments to other sleep)

File: tests/integ/test_stored_procedure.py [hunk 4]
```diff
@@ -1282,6 +1283,297 @@ def select_star_decorator_sp(session_: Session, name: str) -> DataFrame:
...
+        async_job_sproc_table = table_sproc(block=False)
+        while not async_job_sproc_table.is_done():
+            time.sleep(1)
+        result_sproc_table = async_job_sproc_table.result()
+        assert (
+            len(result_sproc_table) == 4
+        ), "StoredProcedure async call should return same data"
+        assert result_sproc_table[1]["TITLE"] == "Movie Two", "Should contain test data"
+
+        # ================ Test 3: None return sproc ================
```
Comment (sfc-gh-aling) [inline tests/integ/test_stored_procedure.py:1393]: is this testing something other than `is_returning_table`?
i'm wondering if test 3 is needed here

File: tests/integ/test_stored_procedure.py (hunk 3 above)
Comment (sfc-gh-mayliu) [reply tests/integ/test_stored_procedure.py]: good catch, removed the `sleep` where applicable. For `test_async_stored_procedure_failure` though, I left `sleep` unchanged since `is_failed` is non-blocking and need completion to return failure

File: tests/integ/test_stored_procedure.py (hunk 4 above)
Comment (sfc-gh-mayliu) [reply tests/integ/test_stored_procedure.py:1393]: IMHO, test 3 provides better coverage with `None` return type for sproc. Async sproc uses either `_AsyncResultType.ROW or _AsyncResultType.COUNT`, so test 3 covers the edge case when sproc doesn't return and verify that `AsyncJob.result()` can empty results well.

File: src/snowflake/snowpark/stored_procedure.py [hunk 5]
```diff
@@ -153,22 +153,29 @@ def __call__(
...
+                statement_params=statement_params,
+                ast_stmt=self._ast_stmt,
+            )
+
+            if not block:
+                return res
+            else:
+                if self._is_return_table:
+                    res._ast = sproc_expr
+                    return res
```
[the remark this answers is not in the document]
Comment (sfc-gh-mayliu) [reply src/snowflake/snowpark/stored_procedure.py]: AST-registered synchronous scalar-returning sproc should wait until `session._ast_batch.eval(self._ast_stmt)` instead of early returning here

Comment (sfc-gh-aling) [inline src/snowflake/snowpark/stored_procedure.py]: do we need this "else" here?
I think lin 181-189 already handles it?
```suggestion
```
